### PR TITLE
net/http: use Copy in ServeContent if CopyN not needed

### PR DESCRIPTION
**Pull request by costela**
This small PR allows optimizations made in io.Copy (like the use of
io.WriterTo) to be used in one possible path of http.ServeContent
(in case of a non-Range request).
This, in turn, allows us to skip the buffer allocation in io.Copy.

This is a retry of CL446276, reverted to address #61530 (whose root
cause has since been addressed)

### CODE REVIEW

**Comment by costela** [comment]
although the [excelent benchmark](https://github.com/golang/go/issues/61530#issuecomment-1646994568) by @AlexanderYastrebov seemed to indicate the underlying issue was solved, the [higher-level benchmark](https://github.com/golang/go/issues/61530#issuecomment-1646701847) by @mauri870 still shows a measurable slowdown (though not quite on the same level as originally observed; tested on current `master`):
```
goos: linux
goarch: amd64
pkg: go61530
cpu: 12th Gen Intel(R) Core(TM) i7-1250U
             │  old.bench  │             new.bench              │
             │   sec/op    │   sec/op     vs base               │
FileServe-12   17.27µ ± 6%   18.12µ ± 2%  +4.94% (p=0.007 n=10)

             │  old.bench   │               new.bench               │
             │     B/op     │     B/op      vs base                 │
FileServe-12   19.97Ki ± 0%   42.73Ki ± 0%  +114.01% (p=0.000 n=10)

             │ old.bench  │             new.bench              │
             │ allocs/op  │ allocs/op   vs base                │
FileServe-12   25.00 ± 0%   28.00 ± 0%  +12.00% (p=0.000 n=10)
```

Since the original [syntetic benchmarks](https://github.com/golang/go/pull/56480#issuecomment-1295451380) remain promising, this looks like another negative interaction with `os.File`.

I'll draft this PR while I investigate further. Sorry for the noise.

**Comment by costela** [comment]
TL;DR: benchmarks were skewed; this PR should cause no performance regressions :tada: :crossed_fingers: 

Longer:

The benchmark results above are skewed by the using `httptest.ResponseRecorder`, which causes a fallback to `os.genericWriteTo` for not being one of the supported targets for a `WriteTo` from an `os.File`. 

Old (`CopyN`)
image

New (`Copy`)
image

The overhead is then caused by missing the [optimization in `copyBuffer` for `LimitReader`](https://github.com/golang/go/blob/7abeefd2b1a03932891e581f1f90656ffebebce4/src/io/io.go#L419). It pretty much disappears as soon as the the response size is bigger than the standard buffer size of 32k:
```
goos: linux
goarch: amd64
pkg: go61530
cpu: 12th Gen Intel(R) Core(TM) i7-1250U
             │  old.bench  │             new.bench              │
             │   sec/op    │   sec/op     vs base               │
FileServe-12   1.425m ± 2%   1.453m ± 5%  +1.95% (p=0.005 n=10)

             │  old.bench   │           new.bench            │
             │     B/op     │     B/op      vs base          │
FileServe-12   75.58Ki ± 3%   76.43Ki ± 3%  ~ (p=0.280 n=10)

             │ old.bench  │             new.bench              │
             │ allocs/op  │ allocs/op   vs base                │
FileServe-12   16.00 ± 0%   19.00 ± 0%  +18.75% (p=0.000 n=10)
```

Adding `ReadFrom` to the `http.ResponseRecorder` also negates the negative effects (I'd say this is an acceptable comparison with `http.response`, which also implements `ReaderFrom`):
```
goos: linux
goarch: amd64
pkg: go61530
cpu: 12th Gen Intel(R) Core(TM) i7-1250U
             │  old.bench  │           new.bench           │
             │   sec/op    │   sec/op     vs base          │
FileServe-12   1.018m ± 7%   1.032m ± 3%  ~ (p=0.165 n=10)

             │  old.bench   │           new.bench            │
             │     B/op     │     B/op      vs base          │
FileServe-12   32.45Ki ± 3%   31.10Ki ± 7%  ~ (p=0.165 n=10)

             │ old.bench  │             new.bench              │
             │ allocs/op  │ allocs/op   vs base                │
FileServe-12   15.00 ± 0%   18.00 ± 0%  +20.00% (p=0.000 n=10)
```

This means this PR shouldn't harm performance for the `io.File` case, and should enable `WriteTo` from sources that support if (e.g. [x/exp/mmap](https://github.com/golang/go/issues/56422#issue-1422973099), hopefully :crossed_fingers:)

**Comment by costela** [comment]
AFAICT the PR should actually _improve_ performance for the realistic `io.File` case because `ReadFrom` in `http.response` will have access to the `io.File` (instead of `io.LimitReader`)? :thinking:

**Comment by costela** [comment]
This still causes a regression in the `io.File` case:
image

1. `io.Copy` tries to use `WriteTo` from the `io.File` (not available if using `io.CopyN`), which fails because `http.response` doesn't implement `syscall.Conn`, so we use `genericWriteTo`
2. that calls a **second** round of `io.Copy`, now with a wrapped `io.File`, so we skip `WriteTo` and go straight to `http.response.ReadFrom`.
3. that then cannot use `sendFile`, because we got a wrapped file as source, so we're back to `genericReadFrom`
4. finally, a **third** round of `io.Copy`, which just falls back to buffered copy.

**Comment by costela** [comment]
With the latest commit the regression is solved. We now see a working `ReadFrom`+`sendFile in the `io.File` case:
image

However, since there are currently multiple proposals open around the intersection of `WriterTo` and `io.Copy` (e.g. #16474, #58331, #56742), it may be necessary to split the last commit into a proposal.

Alternatively, we could also add an internal interface that is only implemented by `io.File`, so that all type checks can be done against it. This would ensure this change has no side-effects outside of the stdlib. In its current form, the change has an unlikely impact, but could still theoretically accidentally match some interface in the wild.
